apis/v1alpha2/referencepolicy_types.go

```diff
@@ -0,0 +1,138 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-api,shortName=refpol
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// ReferencePolicy identifies kinds of resources in other namespaces that are
+// trusted to reference the specified kinds of resources in the same namespace
+// as the policy.
+//
+// Each ReferencePolicy can be used to represent a unique trust relationship.
+// Additional Reference Policies can be used to add to the set of trusted
+// sources of inbound references for the namespace they are defined within.
+//
+// All cross-namespace references in Gateway API (with the exception of cross-namespace
+// Gateway-route attachment) require a ReferencePolicy.
+//
+// Support: Core
+//
+type ReferencePolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of ReferencePolicy.
+	Spec ReferencePolicySpec `json:"spec,omitempty"`
+
+	// Note that `Status` sub-resource has been excluded at the
+	// moment as it was difficult to work out the design.
+	// `Status` sub-resource may be added in future.
+}
+
+// +kubebuilder:object:root=true
+// ReferencePolicyList contains a list of ReferencePolicy.
+type ReferencePolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []ReferencePolicy `json:"items"`
+}
+
+// ReferencePolicySpec identifies a cross namespace relationship that is trusted
+// for Gateway API.
+type ReferencePolicySpec struct {
+	// From describes the trusted namespaces and kinds that can reference the
+	// resources described in "To". Each entry in this list must be considered
+	// to be an additional place that references can be valid from, or to put
+	// this another way, entries must be combined using OR.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	From []ReferencePolicyFrom `json:"from"`
+
+	// To describes the resources that may be referenced by the resources
+	// described in "From". Each entry in this list must be considered to be an
+	// additional place that references can be valid to, or to put this another
+	// way, entries must be combined using OR.
+	//
+	// Support: Core
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	To []ReferencePolicyTo `json:"to"`
+}
+
+// ReferencePolicyFrom describes trusted namespaces and kinds.
+type ReferencePolicyFrom struct {
```

Comment:
I strongly believe that this entire policy work should be separated from gateway. i am guessing the same thing will be used by other apis.

Reply:
I agree that it's likely going to be relevant more broadly. I'd love to see this API move further upstream. I think it's reasonable to test it out here first. For ReferencePolicy to be useful, you have to have a compelling reason to do a cross-namespace reference + an object ref that allows for cross-namespace refs. We have a couple examples of that with this API that seem like a great starting point to vet this concept. I hope it works so well that we can make a case for using it more widely for more APIs.

Reply:
Discussed with sig-auth and consensus was to keep it in Gateway API for now as a POC. If more use cases emerge _and_ they end up being identical, we can explore merging this upstream at that point.